File: src/main/java/com/merge/api/resources/accounting/types/RemoteKey.java
package com.merge.api.resources.accounting.types;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonSetter;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import java.util.Objects;

@JsonInclude(JsonInclude.Include.NON_EMPTY)
@JsonDeserialize(builder = RemoteKey.Builder.class)
public final class RemoteKey {
    private final String name;

    private final String key;

    private RemoteKey(String name, String key) {
        this.name = name;
        this.key = key;
    }

    @JsonProperty("name")
    public String getName() {
        return name;
    }

    @JsonProperty("key")
    public String getKey() {
        return key;
    }

    @Override
    public boolean equals(Object other) {
        if (this == other) return true;
        return other instanceof RemoteKey && equalTo((RemoteKey) other);
    }

    private boolean equalTo(RemoteKey other) {
        return name.equals(other.name) && key.equals(other.key);
    }

    @Override
    public int hashCode() {
        return Objects.hash(this.name, this.key);
    }

    @Override
    public String toString() {
        return "RemoteKey{" + "name: " + name + ", key: " + key + "}";
    }

    public static NameStage builder() {
        return new Builder();
    }

    public interface NameStage {
        KeyStage name(String name);

        Builder from(RemoteKey other);
    }

    public interface KeyStage {
        _FinalStage key(String key);
    }

    public interface _FinalStage {
        RemoteKey build();
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    public static final class Builder implements NameStage, KeyStage, _FinalStage {
        private String name;

        private String key;

        private Builder() {}

        @Override
        public Builder from(RemoteKey other) {
            name(other.getName());
            key(other.getKey());
            return this;
        }

        @Override
        @JsonSetter("name")
        public KeyStage name(String name) {
            this.name = name;
            return this;
        }

        @Override
        @JsonSetter("key")
        public _FinalStage key(String key) {
            this.key = key;
            return this;
        }

        @Override
        public RemoteKey build() {
            return new RemoteKey(name, key);
        }
    }
}

File: merge-java-core/src/main/kotlin/dev/merge/api/models/RemoteKey.kt
package dev.merge.api.models

import com.fasterxml.jackson.annotation.JsonAnyGetter
import com.fasterxml.jackson.annotation.JsonAnySetter
import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import dev.merge.api.core.ExcludeMissing
import dev.merge.api.core.JsonField
import dev.merge.api.core.JsonMissing
import dev.merge.api.core.JsonValue
import dev.merge.api.core.NoAutoDetect
import dev.merge.api.core.toUnmodifiable
import java.util.Objects

/**
 * # The RemoteKey Object
 *
 * ### Description
 *
 * The `RemoteKey` object is used to represent a request for a new remote key.
 *
 * ### Usage Example
 *
 * Post a `GenerateRemoteKey` to receive a new `RemoteKey`.
 */
@JsonDeserialize(builder = RemoteKey.Builder::class)
@NoAutoDetect
class RemoteKey
private constructor(
    private val name: JsonField<String>,
    private val key: JsonField<String>,
    private val additionalProperties: Map<String, JsonValue>,
) {

    private var validated: Boolean = false

    private var hashCode: Int = 0

    fun name(): String = name.getRequired("name")

    fun key(): String = key.getRequired("key")

    @JsonProperty("name") @ExcludeMissing fun _name() = name

    @JsonProperty("key") @ExcludeMissing fun _key() = key

    @JsonAnyGetter
    @ExcludeMissing
    fun _additionalProperties(): Map<String, JsonValue> = additionalProperties

    fun validate(): RemoteKey = apply {
        if (!validated) {
            name()
            key()
            validated = true
        }
    }

    fun toBuilder() = Builder().from(this)

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }

        return other is RemoteKey &&
            this.name == other.name &&
            this.key == other.key &&
            this.additionalProperties == other.additionalProperties
    }

    override fun hashCode(): Int {
        if (hashCode == 0) {
            hashCode =
                Objects.hash(
                    name,
                    key,
                    additionalProperties,
                )
        }
        return hashCode
    }

    override fun toString() =
        "RemoteKey{name=$name, key=$key, additionalProperties=$additionalProperties}"

    companion object {

        @JvmStatic fun builder() = Builder()
    }

    class Builder {

        private var name: JsonField<String> = JsonMissing.of()
        private var key: JsonField<String> = JsonMissing.of()
        private var additionalProperties: MutableMap<String, JsonValue> = mutableMapOf()

        @JvmSynthetic
        internal fun from(remoteKey: RemoteKey) = apply {
            this.name = remoteKey.name
            this.key = remoteKey.key
            additionalProperties(remoteKey.additionalProperties)
        }

        fun name(name: String) = name(JsonField.of(name))

        @JsonProperty("name")
        @ExcludeMissing
        fun name(name: JsonField<String>) = apply { this.name = name }

        fun key(key: String) = key(JsonField.of(key))

        @JsonProperty("key")
        @ExcludeMissing
        fun key(key: JsonField<String>) = apply { this.key = key }

        fun additionalProperties(additionalProperties: Map<String, JsonValue>) = apply {
            this.additionalProperties.clear()
            this.additionalProperties.putAll(additionalProperties)
        }

        @JsonAnySetter
        fun putAdditionalProperty(key: String, value: JsonValue) = apply {
            this.additionalProperties.put(key, value)
        }

        fun putAllAdditionalProperties(additionalProperties: Map<String, JsonValue>) = apply {
            this.additionalProperties.putAll(additionalProperties)
        }

        fun build(): RemoteKey =
            RemoteKey(
                name,
                key,
                additionalProperties.toUnmodifiable(),
            )
    }
}
